### PR TITLE
Quadratic Timelock Delegator

### PR DESCRIPTION
**Pull request by Joeysantoro**
A token timelock for long-term contributors

### CODE REVIEW

File: contracts/utils/timelock/QuadraticTokenTimelock.sol [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+contract QuadraticTokenTimelock is TokenTimelock {
+
+    constructor (
+        address _beneficiary,
+        uint256 _duration,
+        address _lockedToken
+    ) TokenTimelock(
+        _beneficiary, 
+        _duration, 
+        7776000, // 3 months cliff
```
Comment (ditchfieldcaleb) [inline contracts/utils/timelock/QuadraticTokenTimelock.sol:15]: Where's this magic number coming from?

**Review by ditchfieldcaleb — COMMENTED** [review]
Generally looks good, needs more natspec and JS should be TS.

File: contracts/utils/timelock/QuadraticTokenTimelock.sol (hunk 1 above)
Comment (Joeysantoro) [reply contracts/utils/timelock/QuadraticTokenTimelock.sol:15]: its 3 months, in seconds

Comment (eswak) [reply contracts/utils/timelock/QuadraticTokenTimelock.sol:15]: 3*30*24*60*60 to be precise

File: contracts/utils/timelock/QuadraticTokenTimelock.sol [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        _lockedToken, 
+        address(0xB8f482539F2d3Ae2C9ea6076894df36D1f632775) // fei labs multisig
+    ) {}
+
+    function _proportionAvailable(
+        uint256 initialBalance,
+        uint256 elapsed,
+        uint256 duration
+    ) internal pure override returns (uint256) {
+        return initialBalance * elapsed * elapsed / duration / duration;
```
Comment (ElliotFriedman) [inline contracts/utils/timelock/QuadraticTokenTimelock.sol:25]: Why is the formula doing `elapsed * elapsed / duration / duration` ?

File: test/unit/dao/QuadraticTimelock.test.ts [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+          expect(await this.tribe.balanceOf(userAddress)).to.be.bignumber.equal(this.totalTribe);
+        });
+
+        it('updates released amounts', async function () {
+          expect(await this.delegator.alreadyReleasedAmount()).to.be.bignumber.equal(this.totalTribe);
+          expect(await this.delegator.availableForRelease()).to.be.bignumber.equal(toBN(0));
+        });
+      });
+
+      describe.skip('Release To', function () {
```
Comment (ElliotFriedman) [inline test/unit/dao/QuadraticTimelock.test.ts:176]: Remove skip or delete tests

File: test/unit/dao/QuadraticTimelock.test.ts [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+        );
+      });
+      it('Clawback Admin set success', async function () {
+        const clawbackAdmin = await this.delegator.clawbackAdmin();
+        await this.delegator.connect(await getImpersonatedSigner(clawbackAdmin)).clawback();
+      });
+    });
+  });
+
+  describe.skip('Clawback', function () {
```
Comment (ElliotFriedman) [inline test/unit/dao/QuadraticTimelock.test.ts:308]: Remove skip

File: test/unit/dao/QuadraticTimelock.test.ts [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+    }
+  });
+
+  beforeEach(async function () {
+    ({ userAddress, secondUserAddress, beneficiaryAddress1 } = await getAddresses());
+
+    this.tribe = await (await ethers.getContractFactory('MockTribe')).deploy();
+
+    this.window = toBN(4 * 365 * 24 * 60 * 60);
+    this.delegator = await (
```
Comment (ElliotFriedman) [inline test/unit/dao/QuadraticTimelock.test.ts:32]: Remove all references to `this` and instead declare variables and their types globally to this test file.

**Review by ElliotFriedman — CHANGES_REQUESTED** [review]

File: test/unit/dao/QuadraticTimelock.test.ts [hunk 6]
```diff
@@ -0,0 +1,335 @@
...
+        await time.increase((await this.delegator.cliffSeconds()).sub(toBN(1000)));
+        await expectRevert(this.delegator.release(userAddress, '100'), 'TokenTimelock: Cliff not passed');
+      });
+    });
+
+    describe('After cliff', function () {
+      it('releases tokens', async function () {
+        await time.increase((await this.delegator.cliffSeconds()).add(toBN(1)));
+        await this.delegator.release(userAddress, '1');
+        expect(await this.tribe.balanceOf(userAddress)).to.be.bignumber.equal(toBN('1'));
```
Comment (ElliotFriedman) [inline test/unit/dao/QuadraticTimelock.test.ts:65]: Prefer getting starting TRIBE balance, then getting ending TRIBE balance and checking delta.

File: test/unit/dao/QuadraticTimelock.test.ts [hunk 7]
```diff
@@ -0,0 +1,335 @@
+import { expectEvent, expectRevert, getAddresses, getImpersonatedSigner, time } from '../../helpers';
```
Comment (ditchfieldcaleb) [inline test/unit/dao/QuadraticTimelock.test.ts:1]: use shorthand

File: test/unit/dao/QuadraticTimelock.test.ts (hunk 5 above)
Comment (ditchfieldcaleb) [reply test/unit/dao/QuadraticTimelock.test.ts:32]: +1

**Review by ditchfieldcaleb — DISMISSED** [review]
Agreed with Elliots comments as well

**Review by ditchfieldcaleb — APPROVED** [review]
Approved pending test completion

**Pull request merged**